### DISCUSSION TITLE
useAsyncData in my composables

### DISCUSSION THREAD

**Discussion by wokalek** [nuxt/nuxt#21802] [Questions]
Try to use composable in page with module content: 

`composables/blogPosts.ts`:
```ts
export async function useFetchBlogPosts() {
  if (useBlogPosts().value.length) {
    return
  }

  const { data } = await useAsyncData('blogPosts', () => queryContent('/blog').sort({ createdAt: -1, $numeric: false }).find())

  useBlogPosts().value = data.value || []
}
```

`pages/blog/index.vue`
```ts
await useFetchBlogPosts()

definePageMeta({
  colorMode: 'dark',
})
```

And I have this error.

> [nuxt] A composable that requires access to the Nuxt instance was called outside of a plugin, Nuxt hook, Nuxt middleware, or Vue setup function. This is probably not a Nuxt bug. Find out more at `https://nuxt.com/docs/guide/concepts/auto-imports#using-vue-and-nuxt-composables`.

Why I cant use Nuxt composable in my composeble?...

**Comment by baixiaoyu2997** [comment]
![image](https://github.com/nuxt/nuxt/assets/9149999/0aad07ce-7eee-4924-ba61-06eea83c801d)

**Comment by Mini-ghost** [comment]
I'm not sure about the details of `usBlogPosts`, but perhaps you can try the following code:

```ts
export async function useFetchBlogPosts() {
    if (useBlogPosts().value.length) {
        return;
    }

    const nuxtApp = useNuxtApp();

    const { data } = await useAsyncData('blogPosts', () =>
        queryContent('/blog').sort({ createdAt: -1, $numeric: false }).find()
    );

    nuxtApp.runWithContext(() => {
        useBlogPosts().value = data.value || [];
    });
}
```

There is an excellent documentation on `nuxtApp.runWithContext`: https://nuxt.com/docs/api/composables/use-nuxt-app#runwithcontext
